lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Shared.hs

```diff
@@ -0,0 +1,217 @@
+--
+-- An implementation of shared script state using
+-- scheme specified in BIP-0044.
+
+module Cardano.Wallet.Primitive.AddressDiscovery.Shared
+    (
+    -- ** State
+      SharedState (..)
+    , KeyNumberScope (..)
+    , mkSharedState
+    ) where
+
+import Prelude
+
+import Cardano.Address.Script
+    ( Cosigner (..)
+    , KeyHash (..)
+    , Script (..)
+    , ScriptTemplate (..)
+    , ValidationLevel (..)
+    , toScriptHash
+    , validateScriptTemplate
+    )
+import Cardano.Address.Style.Shelley
+    ( Credential (..)
+    , Role (..)
+    , delegationAddress
+    , deriveAddressPublicKey
+    , deriveMultisigPublicKey
+    , getKey
+    , hashKey
+    , liftXPub
+    , mkNetworkDiscriminant
+    )
+import Cardano.Crypto.Wallet
+    ( XPub )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..)
+    , DerivationType (..)
+    , Index (..)
+    , NetworkDiscriminant (..)
+    , WalletKey (..)
+    )
+import Cardano.Wallet.Primitive.AddressDiscovery.Sequential
+    ( DerivationPrefix (..), coinTypeAda, purposeBIP44 )
+import Cardano.Wallet.Primitive.Types.Address
+    ( Address (..) )
+import Control.DeepSeq
+    ( NFData )
+import Control.Monad
+    ( mapM )
+import Data.Either
+    ( isRight )
+import Data.Map.Strict
+    ( Map )
+import Data.Maybe
+    ( isNothing )
+import Data.Word
+    ( Word8 )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Address as CA
+import qualified Data.Map.Strict as Map
+{-------------------------------------------------------------------------------
+                                 State
+-------------------------------------------------------------------------------}
+
+-- | A state to keep track of script templates, account public keys for cosigners,
+-- | and a combination of possible script addresses to look for in a ledger
+data SharedState (n :: NetworkDiscriminant) k = SharedState
+    { accountKey :: k 'AccountK XPub
+        -- ^ Reward account public key associated with this wallet
+    , derivationPrefix :: DerivationPrefix
+        -- ^ Derivation path prefix from a root key up to the account key
+    , derivationKeyNumber :: KeyNumberScope
+        -- ^ Number of first keys that are used to produce candidate addresses
+    , paymentScriptTemplate :: !(Maybe ScriptTemplate)
+        -- ^ Script template together with a map of account keys and cosigners
+        -- for payment credential
+    , stakingScriptTemplate :: !(Maybe ScriptTemplate)
+        -- ^ Script template together with a map of account keys and cosigners
+        -- for staking credential
+    , addressCandidates :: [Address]
+    }
+    deriving stock (Generic)
+
+deriving instance
+    ( Show (k 'AccountK XPub)
+    ) => Show (SharedState n k)
+
+deriving instance
+    ( Eq (k 'AccountK XPub)
+    ) => Eq (SharedState n k)
+
+instance
+    ( NFData (k 'AccountK XPub)
+    )
+    => NFData (SharedState n k)
+
+newtype KeyNumberScope =
+    KeyNumberScope { unKeyNumberScope :: Word8 }
```

Comment:
I am not sure whether I understand the purpose of the `KeyNumberScope` fully yet, but a `Word8` seems small?